File: modules/gitwcsub/files/app/gitwcsub.lua
local cfgFile = "gitwcsub.cfg" -- change this to the location of the config file

local socket = require"socket" -- LuaSocket: http://files.luaforge.net/releases/luasocket/luasocket/luasocket-2.0.2
local JSON = require "JSON" -- JSON: http://regex.info/code/JSON.lua
local config = require "config"

local cfg = config.read(cfgFile)
local lastLogRotate = 0

local retries, lastping, lastConfigRead = 0, 0, os.time()
local _print = print
local logFD;

local function print(txt)
    local now = os.time()
    if not logFD then
        logFD = io.open(cfg.logging.logFile, "w+")
        if not logFD then 
            _print("Could not open log file " .. cfg.logging.logFile .. " for writing!!")
            os.exit()
        end
    end
    if cfg and cfg.logging.logFile then
        -- First, check if midnight has passed and we need to create a new log.
        if (now % cfg.logging.rotateTime) <= 20 and now > (lastLogRotate+20) then
            if logRotate ~= 0 then
                _print("Rotating logs...")
                logFD:close()
            end
            local newName = os.date("!" .. cfg.logging.logArchives)
            _print("Relocating "..cfg.logging.logFile .. " to " .. newName)
            os.execute( ("mv %s %s"):format(cfg.logging.logFile, newName) )
            local ancientName = os.date("!" .. cfg.logging.logArchives, now - (cfg.logging.logLimit * cfg.logging.rotateTime) )
            os.execute( ("rm %s"):format(ancientName) )
            lastLogRotate = now
            logFD = io.open(cfg.logging.logFile, "w+")
        end
        logFD:write(os.date("![%Y-%m-%d %h:%M:%S]: ") .. txt .. "\n")
        logFD:flush()
        _print (os.date("![%Y-%m-%d %h:%M:%S]: ") .. txt)
    else
        _print(txt)
    end
end
        

function connectToPubSub(url)
    local server, port, uri = url:match("^([^:]+):(%d+)(/.+)$")
    print("Connecting to PubSub server " .. server)
    local s = socket.tcp()
    s:settimeout(2)
    local success, err = s:connect(socket.dns.toip(server) or server, tonumber(port))
    if not success then
        print("Failed to connect: ".. err)
        return false
    end
    s:send("GET " .. uri .. " HTTP/1.1\r\n");
    s:send("Host: " .. server .. "\r\n\r\n");
    s:settimeout(20)
    return s
end

-- updateRepository: Updates a repository or clones a new one if it doesn't exist.
function updateRepository(path, repository)
    -- Check that the repo exists
    local f = io.popen("ls " .. path)
    if f then
        local data = f:read("*l")
        f:close()
        local pullFailed = false
        if data then
            -- if the repository does exist, pull from it
            print( ("Pulling from %s%s.git into %s"):format(cfg.servers.git, repository, path) )
            local rc, foo, rv = os.execute( ("git --git-dir %s/.git pull"):format(path) )
            if rv ~= 0 then
                pullFailed = true
                print( ("Pull from %s%s.git into %s failed, trying clone instead"):format(cfg.servers.git, repository, path) )
            end
        end
        if not data or pullFailed then
            -- if it doesn't exist, or if pull failed (folder exists but is empty), create it by cloning
            print( ("Cloning from %s%s.git into %s"):format(cfg.servers.git, repository, path) )
            local rc, foo, rv = os.execute( ("git clone -b %s --single-branch %s%s.git %s"):format(cfg.misc.refreal, cfg.servers.git, repository, path) )
            if rv ~= 0 then
                print( ("Could not clone from %s%s.git into %s!"):format(cfg.servers.git, repository, path) )
            end
        end
    end
end


-- readPubSub: Tries to read a line from the pubsub
function readPubSub(s)
    lastping = os.time()
    while true do
    
        -- Reread the configuration every 30 seconds, so we won't have to restart anything
        if (lastConfigRead + 30) < os.time() then
            if cfg.misc.debug then
                print("Reloading configuration...")
            end
            lastConfigRead = os.time()
            cfg = config.read(cfgFile)
        end
        
        -- Try to fetch a json object from pubsub
        local receive, err = s:receive('*l')
        if receive and receive:match("^([[a-zA-Z0-9]+)$") then
            local howMuch = tonumber(receive, 16)
            receive, err = s:receive(howMuch+2)
        end
        
        -- We got an object, parse it and see if it matches a site object
        if receive then
            if cfg.misc.debug then
                print(receive)
            end
            lastping = os.time()
            if receive:match([[^{%s*"commit"%s*:]]) and receive:len() > 3 then
                local line = receive:gsub("\0", ""):gsub(",\r?\n", "")
                local okay, json = pcall(function() return JSON:decode(line) end)
                if okay and json then
                    local commit = json.commit or {}
                    if commit.log and commit.repository and commit.repository == "git" then
                        if cfg.misc.debug then
                            print("Got git push from repository " .. (commit.project or "?") .. ".git")
                        end
                        -- For each repository...
                        for path, repository in pairs(cfg.tracking) do
                            -- Do we have a match?
                            if commit.project and commit.project == repository then
                                if commit.ref == cfg.misc.ref then
                                    if cfg.misc.debug then
                                        print( ("Pulling from %s.git"):format(commit.project) )
                                    end
                                    -- We have a match, let's update the local copy (or create it)
                                    updateRepository(path, repository)
                                end
                            end
                        end
                    end
                end
            end
        -- various timeouts
        else
            if err == "timeout" then
                if lastping < (os.time() - 30) then
                    err = "disconnected (ping timeout)"
                    retries = 0
                end
            end
            if err ~= "timeout" then
                print("Connection closed: " .. err)
                s = connectToPubSub(cfg.servers.pubsub)  -- reconnect to PubSub server
                if s then 
                    retries = 0
                else
                    retries = retries + 1
                end
            end
            
            if not s and retries > 100 then
                print("Disconnected and could not reconnect, giving up.")
                break -- timeout, nothing more to receive at the moment, let's get back to responding to PINGs
            end
        end
    end
end


--[[
Main program
]]
print("Program started, updating/populating repositories:")
for path, repository in pairs(cfg.tracking) do
    print("Checking " .. path .. "...")
    updateRepository(path, repository)
end

local s = connectToPubSub(cfg.servers.pubsub)
if s then
    readPubSub(s)
else
    print("Failed to connect to pubsub service!")
end

print("Bye for now!")

File: modules/gitwcsub/files/app/gitwcsub.py
from threading import Thread
import json, urllib.request, urllib.parse, configparser, re, base64, sys, os, time, atexit, signal, logging, subprocess, collections, argparse, grp, pwd, shutil
from threading import Lock
import threading;
from collections import namedtuple
import random, atexit, signal, inspect
import time
import logging
import logging.handlers


path = os.path.dirname(os.path.abspath(sys.argv[0]))
if len(path) == 0:
    path = "."
    
# Fetch config
config = configparser.RawConfigParser()

config.read(path + '/gitwcsub.cfg')



log_handler = logging.handlers.WatchedFileHandler(config.get("Logging", "logFile"))
formatter = logging.Formatter(
	'%(asctime)s gitwcsub [%(process)d]: %(message)s',
	'%b %d %H:%M:%S')
formatter.converter = time.gmtime
log_handler.setFormatter(formatter)
logger = logging.getLogger()
logger.addHandler(log_handler)
logger.setLevel(logging.INFO)

pending = {}



class Daemonize:
	"""A generic daemon class.

	Usage: subclass the daemon class and override the run() method."""

	def __init__(self, pidfile): self.pidfile = pidfile
	
	def daemonize(self):
		"""Deamonize class. UNIX double fork mechanism."""

		try: 
			pid = os.fork() 
			if pid > 0:
				# exit first parent
				sys.exit(0) 
		except OSError as err: 
			sys.stderr.write('fork #1 failed: {0}\n'.format(err))
			sys.exit(1)
	
		# decouple from parent environment
		os.chdir('/') 
		os.setsid() 
		os.umask(0) 
	
		# do second fork
		try: 
			pid = os.fork() 
			if pid > 0:

				# exit from second parent
				sys.exit(0) 
		except OSError as err: 
			sys.stderr.write('fork #2 failed: {0}\n'.format(err))
			sys.exit(1) 
	
		# redirect standard file descriptors
		sys.stdout.flush()
		sys.stderr.flush()
		si = open(os.devnull, 'r')
		so = open(os.devnull, 'a+')
		se = open(os.devnull, 'a+')

		os.dup2(si.fileno(), sys.stdin.fileno())
		os.dup2(so.fileno(), sys.stdout.fileno())
		os.dup2(se.fileno(), sys.stderr.fileno())
	
		# write pidfile
		atexit.register(self.delpid)

		pid = str(os.getpid())
		with open(self.pidfile,'w+') as f:
			f.write(pid + '\n')
	
	def delpid(self):
		os.remove(self.pidfile)

	def start(self, args):
		"""Start the daemon."""

		# Check for a pidfile to see if the daemon already runs
		try:
			with open(self.pidfile,'r') as pf:

				pid = int(pf.read().strip())
		except IOError:
			pid = None
	
		if pid:
			message = "pidfile {0} already exist. " + \
					"Daemon already running?\n"
			sys.stderr.write(message.format(self.pidfile))
			sys.exit(1)
		
		# Start the daemon
		self.daemonize()
		self.run(args)

	def stop(self):
		"""Stop the daemon."""

		# Get the pid from the pidfile
		try:
			with open(self.pidfile,'r') as pf:
				pid = int(pf.read().strip())
		except IOError:
			pid = None
	
		if not pid:
			message = "pidfile {0} does not exist. " + \
					"Daemon not running?\n"
			sys.stderr.write(message.format(self.pidfile))
			return # not an error in a restart

		# Try killing the daemon process	
		try:
			while 1:
				os.kill(pid, signal.SIGTERM)
				time.sleep(0.1)
		except OSError as err:
			e = str(err.args)
			if e.find("No such process") > 0:
				if os.path.exists(self.pidfile):
					os.remove(self.pidfile)
			else:
				print (str(err.args))
				sys.exit(1)

	def restart(self):
		"""Restart the daemon."""
		self.stop()
		self.start()

	def run(self):
		"""You should override this method when you subclass Daemon.
		
		It will be called after the process has been daemonized by 
		start() or restart()."""


# Func for cloning a new repo
def buildRepo(path, URL, branch):
    logging.info("%s does not exist, trying to clone into it as a new dir" % path)
    rv = "No output"
    try:
        rv = subprocess.check_output(("git", "clone", "-b", branch, "--single-branch", URL, path))
    except Exception as err:
        rv = "Error while cloning: %s" % err
    logging.info(rv)
   
   
# Func for updating whatever is in the queue
def updatePending():
    global pending
    xpending = pending;
    pending = {}
    for entry in xpending:
        repo = entry
        path = xpending[entry]
        trueRepo = repo
        httproot = config.get("Servers", "gitroot")
        # Check if non-default git root, if so find new http root
        m = re.match(r"^\$([^/]+)/(.+)$", repo)
        if m:
            trueRepo = m.group(2)
            groot = m.group(1)
            httproot = config.get("Servers", groot) if config.has_option("Servers", groot) else httproot
        # Special branch via repo:branch syntax? Otherwise, default to config global
        branch = config.get("Misc", "branch")
        if ':' in trueRepo:
            trueRepo, branch = trueRepo.split(':', 1)
        newURL = "%s%s.git" % (httproot, trueRepo)
        
        # Check if we need to pull or clone:
        
        # Existing repo?
        if os.path.isdir(path):
            os.chdir(path)
            
            # Check if the repo has moved, and if so, tear down and rebuild
            try:
                oldURL = ""
                try:
                    oldURL = str( subprocess.check_output(("git", "config", "remote.origin.url")), encoding='ascii' ).rstrip('\r\n')
                except Exception as err:
                    logging.warn("Git config exited badly, not a Git repo??")
                    oldURL = ""
                if newURL != oldURL:
                    canClobber = True
                    if oldURL == "":
                        logging.warn("No previous git remote detected, checking for SVN info..")
                        canClobber = False
                        try:
                            rv =subprocess.check_output(("svn", "info"))
                            if re.search(r"URL: ", rv.decode("utf-8"), re.MULTILINE):
                                canClobber = True
                                logging.warn("This is/was a subversion repo, okay to clobber")
                        except Exception as err:
                            logging.warn("Not a repo, not going to clobber it: %s" % err)
                            canClobber = False
                    if canClobber:
                        logging.warn("Local origin (%s) does not match configuration origin (%s), rebuilding!" % (oldURL, newURL))
                        os.chdir("/")
                        shutil.rmtree(path)
                        buildRepo(path, newURL, branch)
            # Otherwise, just pull in changes
                else:
                    logging.info("Pulling changes into %s" % path)
                    rv = subprocess.check_output(("git", "fetch", "origin", branch))
                    logging.info(rv)
                    rv = subprocess.check_output(("git", "reset", "--hard", "origin/%s" % branch))
                    logging.info(rv)
            except Exception as err:
                logging.error("Git update error: %s" % err)
                
        # New repo?
        else:
            buildRepo(path, newURL, branch)
            
def read_chunk(req):
    while True:
        try:
            line = req.readline().strip()
            if line:
                yield line
            else:
                print("No more lines?")
                break
        except Exception as info:
            logging.warning("Error reading from stream: %s", info)
            break
    return
    
def parseGitCommit(commit):
    global pending
    if commit['repository'] == "git":
        if 'project' in commit:
            project = commit['project']
            branch = commit['ref'].replace("refs/heads/", "")
            for option in config.options("Tracking"):
                branchNeeded = config.get("Misc", "branch")
                path = option
                repo = config.get("Tracking", option)
                trueRepo = repo
                # Check if non-default git root, weed that out if so
                m = re.match(r"^\$([^/]+)/(.+)$", repo)
                if m:
                    trueRepo = m.group(2)
                    groot = m.group(1)
                # Check for non-standard publish branch
                if ':' in trueRepo:
                    trueRepo, branchNeeded = trueRepo.split(':', 1)
                if project == trueRepo and branch == branchNeeded:
                    logging.info("Adding %s (%s) to the update queue" % (path, repo))
                    pending[repo] = path
                

            
    
# PubSub class: handles connecting to a pubsub service and checking commits
class PubSubClient(Thread):
   
    
    def run(self):
        self.killed = False
        while not self.killed:
            logging.info("[%s] Connecting to " % self.ident + self.url + "...")
            self.req = None
            while not self.req:
                try:
                    self.req = urllib.request.urlopen(self.url, None, 30)
                    logging.info("[%s] Connected to " % self.ident + self.url + ", reading stream")
                except:
                    logging.warning("[%s] Could not connect to %s, retrying in 30 seconds..." % (self.ident, self.url))
                    time.sleep(30)
                    continue
                
            for line in read_chunk(self.req):
                line = str( line, encoding='ascii', errors='ignore' ).rstrip('\r\n,').replace('\x00','') # strip away any old pre-0.9 commas from gitpubsub chunks and \0 in svnpubsub chunks
                
                try:
                    obj = json.loads(line)
                    if "commit" in obj and "repository" in obj['commit']:
                        logging.info("[%s] Got a commit in '%s'" % (self.ident, obj['commit']['repository']))
                        if obj['commit']['repository'] == "git":
                            parseGitCommit(obj['commit'])
                            
                except ValueError as detail:
                    logging.warning("Bad JSON or something: %s", detail)
                    continue
            logging.info("[%s] Disconnected from %s, reconnecting" % (self.ident, self.url))
            if self.killed:
               logging.info("[%s] Pubsub thread killed, exiting" % self.ident)
               break

parser = argparse.ArgumentParser(description='Command line options.')
parser.add_argument('--user', dest='user', type=str, nargs=1,
                   help='Optional user to run GitWcSub as')
parser.add_argument('--group', dest='group', type=str, nargs=1,
                   help='Optional group to run GitWcSub as')
parser.add_argument('--pidfile', dest='pidfile', type=str, nargs=1,
                   help='Optional pid file location')
parser.add_argument('--daemonize', dest='daemon', action='store_true',
                   help='Run as a daemon')
parser.add_argument('--stop', dest='kill', action='store_true',
                   help='Kill the currently running GitWcSub process')
args = parser.parse_args()

pidfile = "/var/run/gitwcsub.pid"
if args.pidfile and len(args.pidfile) > 2:
    pidfile = args.pidfile

def main():
    
    if args.group and len(args.group) > 0:
        gid = grp.getgrnam(args.group[0])
        os.setgid(gid[2])
    
    if args.user and len(args.user) > 0:
        print("Switching to user %s" % args.user[0])
        uid = pwd.getpwnam(args.user[0])[2]
        os.setuid(uid)
    
    global pending
    pubsub = PubSubClient()
    pubsub.url = config.get("Servers", "pubsub")
    pubsub.start()
    
    logging.warn("Service restarted, checking for updates in all tracked repos")
    for option in config.options("Tracking"):
        repo = config.get("Tracking", option)
        path = option
        pending[repo] = path
    
    while True:
        updatePending()
        time.sleep(3)
        
    
## Daemon class
class MyDaemon(Daemonize):
    def run(self, args):
        main()
    
# Get started!
if args.kill:
    print("Stopping GitWcSub")
    daemon = MyDaemon(pidfile)
    daemon.stop()
else:
    if args.daemon:
        print("Daemonizing...")
        daemon = MyDaemon(pidfile)
        daemon.start(args)
    else:
        main()
